modin/engines/modin_dataframe.py

```diff
@@ -0,0 +1,406 @@
+                axis.
+        """
+        pass
+
+    def groupby(self, axis, by, operator, result_schema=None):
+        """Generates groups based on values in the input column(s) and performs
+        the specified operation (e.g. aggregation or backfill) on the groups.
+
+        Notes
+        -----
+            No communication between groups is allowed in this algebra implementation.
+
+            The number of rows (columns if axis=1) returned by the user-defined function
+                passed to the groupby may be at most the number of rows in the group, and
+                may be as small as a single row.
+
+            Unlike the pandas API, an intermediate “GROUP BY” object is not present in this
+                algebra implementation.
+
+        Parameters
+        ----------
+            axis: int
+                The axis to apply the grouping over.
+            by: string or list of strings
+                One or more column labels to use for grouping.
+            operator: callable
+                The operation to carry out on each of the groups. The operator is another
+                algebraic operator with its own user-defined function parameter, depending
+                on the output desired by the user.
+            result_schema: list of dtypes
+                List of data types that represent the types of the output dataframe.
+
+        Returns
+        -------
+        ModinDataframe
+            A new ModinDataframe containing the groupings specified, with the operator
+                applied to each group.
+        """
+        pass
+
+    def reduction(self, axis, function, tree_reduce=False, result_schema=None):
+        """Performs a user-defined per-column aggregation, where each column reduces
+        down to a single value.
+
+        Notes
+        -----
+            The user-defined function must reduce to a single value.
+
+        Parameters
+        ----------
+            axis: int
+                The axis to perform the reduction over.
+            function: callable
+                The reduction function to apply to each column.
+            tree_reduce: boolean
+                Flag to signal to the compiler that the function
+                can be applied using a tree reduction (e.g. max or sum).
+                Set this flag to False for functions that need to look at
+                the entire column at once to perform their reduction (e.g. median).
+            result_schema: list of dtypes
+                List of data types that represent the types of the output dataframe.
+
+        Returns
+        -------
+        ModinDataframe
+            A new ModinDataframe with the same columns as the previous, with only a single row.
+        """
+        pass
+
+    def infer_types(self, columns_list):
+        """Determines the compatible type shared by all values in the specified columns,
+        and converts all values to that type.
+
```

Comment:
Do we want to include a discussion of when exceptions are thrown here?

Reply:
Yes, it should just be Python TypeErrors in this case. Any others?

Reply:
This was meant to be a continuation of the discussion stemming from the section of the paper that stated that infer_types would throw an error - you mentioned that if the types couldn't be coerced, an error should be thrown; however, depending on the type system, no errors may be thrown?

Reply:
I don't think we're adding rules for coercing types. The `TypeError` would probably not happen in this method.